apollo-federation/src/sources/connect/json_selection/location.rs

```diff
@@ -0,0 +1,372 @@
+// produce a range directly from their children, so they do not need to be
+// wrapped as WithRange<PathSelection> or WithRange<NamedSelection>.
+// Additionally, AST nodes that are structs can store their own range as a
+// field, so they can implement Ranged<T> without the WithRange<T> wrapper.
+pub trait Ranged<T> {
+    fn node(&self) -> &T;
+    fn range(&self) -> OffsetRange;
+}
+
+// The ranges produced by the JSONSelection parser are pairs of character
+// offsets into the original string. The first element of the pair is the offset
+// of the first character, and the second element is the offset of the character
+// just past the end of the range. Offsets start at 0 for the first character in
+// the file, following nom_locate's span.location_offset() convention.
+pub type OffsetRange = Option<(usize, usize)>;
+
+// The most common implementation of the Ranged<T> trait is the WithRange<T>
+// struct, used to wrap any AST node that (a) needs its own location information
+// (because that information is not derivable from its children) and (b) cannot
+// easily store that information by adding another struct field (most often
+// because T is an enum or primitive/String type, not a struct).
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct WithRange<T> {
+    node: Box<T>,
+    range: OffsetRange,
+}
+
+// We can recover some of the ergonomics of working with the inner type T by
+// implementing Deref and DerefMut for WithRange<T>.
+impl<T> std::ops::Deref for WithRange<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.node.as_ref()
+    }
+}
+impl<T> std::ops::DerefMut for WithRange<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.node.as_mut()
+    }
+}
+
+impl<T> AsRef<T> for WithRange<T> {
+    fn as_ref(&self) -> &T {
+        self.node.as_ref()
+    }
+}
+
+impl<T> AsMut<T> for WithRange<T> {
+    fn as_mut(&mut self) -> &mut T {
+        self.node.as_mut()
+    }
+}
+
+impl<T> PartialEq<T> for WithRange<T>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &T) -> bool {
+        self.node.as_ref() == other
+    }
+}
+
+impl<T> Ranged<T> for WithRange<T> {
+    fn node(&self) -> &T {
+        self.node.as_ref()
+    }
+
+    fn range(&self) -> OffsetRange {
+        self.range
+    }
+}
+
+impl<T> WithRange<T> {
+    pub(crate) fn new(node: T, range: OffsetRange) -> Self {
+        Self {
+            node: Box::new(node),
+            range,
+        }
+    }
+
+    pub(crate) fn take(self) -> T {
+        *self.node
+    }
+
+    pub(crate) fn take_as<U>(self, f: impl FnOnce(T) -> U) -> WithRange<U> {
+        WithRange::new(f(*self.node), self.range)
+    }
+}
+
+pub(super) fn merge_ranges(left: OffsetRange, right: OffsetRange) -> OffsetRange {
+    match (left, right) {
+        (Some((left_start, _left_end)), Some((_right_start, right_end))) => {
+            Some((left_start, right_end))
+        }
```

Comment:
What if ranges overlap or aren't in increasing order? Might be useful to have a guard here:

```suggestion
        (Some((left_start, _left_end)), Some((right_start, right_end))) if left_start <  right_start => {
            Some((left_start, right_end))
        }
```

Or something similar to ensure that this range is never invalid

Reply:
Included in https://github.com/apollographql/router/pull/5987/commits/774ae142dcb59a8699bf1c685a2d48c00bf56311.

Reply:
What should happen if the ranges aren't continuous e.g. 1..3 and 9..10. Should they be merged to 1..10?

Reply:
Yes, often the children of a given AST node will be separated by spaces/comments, so they won't be contiguous, but the parent range should encompass all the child ranges.